src/Storages/MergeTree/IMergeTreeDataPart.cpp

```diff
@@ -1325,6 +1338,91 @@ String IMergeTreeDataPart::getRelativePathForDetachedPart(const String & prefix)
+            part_virtual->parent_part = this;
+            projection_parts.emplace(ProjectionDescription::VIRTUAL_PROJECTION_NAME, std::move(part_virtual));
+            return;
+        }
+        else
+            throw Exception(
+                "Cannot add virtual projection part when minmax_idx is not initialized yet. It's a bug", ErrorCodes::LOGICAL_ERROR);
+    }
+
+    auto minmax_count_columns = block.mutateColumns();
+    size_t minmax_idx_size = minmax_idx.hyperrectangle.size();
+    if (2 * minmax_idx_size != minmax_count_columns.size() - 1)
+        throw Exception(
+            ErrorCodes::LOGICAL_ERROR,
+            "Virtual projection part should have twice plus one the number of ranges in minmax_idx. 2 * minmax_idx_size = {}, "
+            "minmax_count_columns.size() - 1 = {}. It's a bug",
+            2 * minmax_idx_size,
+            minmax_count_columns.size() - 1);
+
+    auto insert = [&](ColumnAggregateFunction & column, const Field & value)
+    {
+        auto func = column.getAggregateFunction();
+        Arena & arena = column.createOrGetArena();
+        size_t size_of_state = func->sizeOfData();
+        size_t align_of_state = func->alignOfData();
+        auto * place = arena.alignedAlloc(size_of_state, align_of_state);
+        func->create(place);
+        auto value_column = func->getReturnType()->createColumnConst(1, value)->convertToFullColumnIfConst();
+        const auto * value_column_ptr = value_column.get();
+        func->add(place, &value_column_ptr, 0, &arena);
+        column.insertFrom(place);
+    };
+
+    for (size_t i = 0; i < minmax_idx_size; ++i)
+    {
+        size_t min_pos = i * 2;
+        size_t max_pos = i * 2 + 1;
+        auto & min_column = assert_cast<ColumnAggregateFunction &>(*minmax_count_columns[min_pos]);
+        auto & max_column = assert_cast<ColumnAggregateFunction &>(*minmax_count_columns[max_pos]);
+        auto & range = minmax_idx.hyperrectangle[i];
+        insert(min_column, range.left);
+        insert(max_column, range.right);
+    }
+
+    {
+        auto & column = assert_cast<ColumnAggregateFunction &>(*minmax_count_columns.back());
+        auto func = column.getAggregateFunction();
+        Arena & arena = column.createOrGetArena();
+        size_t size_of_state = func->sizeOfData();
+        size_t align_of_state = func->alignOfData();
+        auto * place = arena.alignedAlloc(size_of_state, align_of_state);
+        func->create(place);
+        const AggregateFunctionCount & agg_count = assert_cast<const AggregateFunctionCount &>(*func);
+        agg_count.set(place, rows_count);
+        column.insertFrom(place);
+    }
+    block.setColumns(std::move(minmax_count_columns));
+    auto part_virtual = std::make_shared<MergeTreeDataPartInMemory>(storage, block);
+    part_virtual->parent_part = this;
+    projection_parts.emplace(ProjectionDescription::VIRTUAL_PROJECTION_NAME, std::move(part_virtual));
```

Comment:
So, we store this projection as in-memory part.
I think that creating it at read time may be better